File: LinearSystem/src/GaussianElimination.cpp
#include<iostream>
#include<cmath>
#include "GaussianElimination.h"
using namespace std;

GaussianElimination::GaussianElimination(int sz) : LinearSystem(sz)
{}

vector<double>partialPivot::solve()
{
    forwardElimination();
    return backSubstitution();
}

vector<double> withoutPivot::solve()
{
    forwardElimination();
    return backSubstitution();
}

void GaussianElimination::forwardElimination()
{
    for(int k=0;k<rows;k++)
    {
        pivot(k);
        if(fabs(a[k][k]) < 1e-12)
            throw "zero pivot";
        for(int i=k+1;i<rows;i++)
        {
            double factor=a[i][k]/a[k][k];
            for(int j=k;j<cols;j++)
                a[i][j]-=factor*a[k][j];
        }

    }
}

vector<double> GaussianElimination::backSubstitution()
{
    vector <double>x(rows);
    for(int i=rows-1;i>=0;i--)
    {
        x[i]=a[i][cols-1];
        for(int j=i+1;j<rows;j++)
            x[i]-=a[i][j]*x[j];
        x[i]/=a[i][i];
    }
    return x;
}

withoutPivot::withoutPivot(int sz):GaussianElimination(sz)
{}

void withoutPivot::pivot(int k)
{}

partialPivot::partialPivot(int sz):GaussianElimination(sz)
{}

void partialPivot::pivot(int k)
{
    int maxRow=k;
    for(int i=k+1;i<rows;i++)
    {
        if(fabs(a[i][k])>fabs(a[maxRow][k]))
            maxRow=i;
    }
    if(maxRow!=k)
        swap(a[k],a[maxRow]);
}

File: LinearSystem in Python/src/GaussianElimination.py
import math
from LinearSystem import LinearSystem


class GaussianElimination(LinearSystem):
    def __init__(self, sz):
        super().__init__(sz)

    def pivot(self, k):
        raise NotImplementedError

    def forward_elimination(self):
        for k in range(self.rows):
            self.pivot(k)
            if abs(self.a[k][k]) < 1e-12:
                raise Exception("zero pivot")
            for i in range(k + 1, self.rows):
                factor = self.a[i][k] / self.a[k][k]
                for j in range(k, self.cols):
                    self.a[i][j] -= factor * self.a[k][j]

    def back_substitution(self):
        x = [0.0] * self.rows
        for i in range(self.rows - 1, -1, -1):
            x[i] = self.a[i][self.cols - 1]
            for j in range(i + 1, self.rows):
                x[i] -= self.a[i][j] * x[j]
            x[i] /= self.a[i][i]
        return x


class WithoutPivot(GaussianElimination):
    def __init__(self, sz):
        super().__init__(sz)

    def pivot(self, k):
        pass

    def solve(self):
        self.forward_elimination()
        return self.back_substitution()


class PartialPivot(GaussianElimination):
    def __init__(self, sz):
        super().__init__(sz)

    def pivot(self, k):
        max_row = k
        for i in range(k + 1, self.rows):
            if abs(self.a[i][k]) > abs(self.a[max_row][k]):
                max_row = i
        if max_row != k:
            self.a[k], self.a[max_row] = self.a[max_row], self.a[k]

    def solve(self):
        self.forward_elimination()
        return self.back_substitution()
